src/v/pandaproxy/schema_registry/seq_writer.cc

```diff
@@ -71,110 +71,299 @@ ss::future<> seq_writer::read_sync_inner() {
-    // we'll just retry, rather than writing invalid stuff.
-    auto projected = co_await _store.project_ids(sub, def, type);
-
-    if (!projected.inserted) {
-        vlog(plog.debug, "write_subject_version: no-op");
+/// Permanent deletions (i.e. writing tombstones for previous sequenced
+/// records) do not themselves need sequence numbers.
+/// Include a version if we are only to hard delete that version, otherwise
+/// will hard-delete the whole subject.
+ss::future<std::vector<schema_version>> seq_writer::delete_subject_permanent(
+  subject sub, std::optional<schema_version> version) {
+    std::vector<seq_marker> sequences;
+    /// Check for whether our victim is already soft-deleted happens
+    /// within these store functions (will throw a 404-equivalent if so)
+    vlog(plog.debug, "delete_subject_permanent sub={}", sub);
+    if (version.has_value()) {
+        sequences = co_await _store.get_subject_version_written_at(
+          sub, version.value());
     } else {
+        sequences = co_await _store.get_subject_written_at(sub);
+    }
+
+    storage::record_batch_builder rb{
+      model::record_batch_type::raft_data, model::offset{0}};
+
+    std::vector<std::variant<schema_key, delete_subject_key>> keys;
+    for (auto s : sequences) {
         vlog(
           plog.debug,
-          "seq_writer::write_subject_version project offset={} subject={} "
-          "schema={} "
-          "version={}",
-          write_at,
+          "Delete subject_permanent: tombstoning sub={} at {}",
           sub,
-          projected.id,
-          projected.version);
+          s);
 
-        auto my_node_id = config::shard_local_cfg().node_id();
+        // FIXME: assuming magic is the same as it was when key was
+        // originally read... remove magic field now that we aren't aiming
+        // for topic-level compatibility?
+        if (s.delete_subject) {
+            auto key = delete_subject_key{
+              .seq{s.seq}, .node{s.node}, .sub{sub}};
+            keys.push_back(key);
+            rb.add_raw_kv(to_json_iobuf(std::move(key)), std::nullopt);
+        } else {
+            auto key = schema_key{
+              .seq{s.seq}, .node{s.node}, .sub{sub}, .version{s.version}};
+            keys.push_back(key);
+            rb.add_raw_kv(to_json_iobuf(std::move(key)), std::nullopt);
+        }
+    }
 
-        auto key = schema_key{
-          .seq{write_at},
-          .node{my_node_id},
-          .sub{sub},
-          .version{projected.version}};
-        auto value = schema_value{
-          .sub{std::move(sub)},
-          .version{projected.version},
-          .type = type,
-          .id{projected.id},
-          .schema{std::move(def)},
-          .deleted = is_deleted::no};
+    // If a subject is in the store, it must have been replayed some somewhere,
+    // so there must be some entries in the list of keys to tombstone.
+    assert(!keys.empty());
```

Comment:
We tend not to use `assert`. Can it be handled gracefully at runtime, are you trying to spot a logic error? For logic errors we tend to use vassert, which segfaults - that way we get a nice stacktrace and prevent (further) data corruption.